### PR TITLE
[FLINK-38835][postgres] Fix timestamp conversion to LocalDateTime for dates before 1970-01-01.

### PR DESCRIPTION
**Pull request by lvyanquan**
Time offsets before and after 1970-01-01 may be inconsistent, which could result in different data being returned during the full phase and the incremental phase. 
Here, the conversion to LocalDateTime is performed using Timestamp.toLocalDateTime to bypass timezone issues.

### CODE REVIEW

**Comment by yuxiqian** [comment]
Could you please resolve the conflicts?

**Review by leonardBang — COMMENTED** [review]
## Review: [FLINK-38835][postgres] Fix timestamp conversion to LocalDateTime for dates before 1970-01-01

Reviewed against the diff (+116/-6, 5 files), Debezium `1.9.8.Final`, pgjdbc `42.7.3`. Cross-checked `PostgresValueConverter.convertTimestampToLocalDateTime` (decompiled) and measured old vs. new behavior under `Asia/Shanghai`, `Europe/Paris`, `UTC`, and `GMT±hh:mm`.

### Strengths

- **Root cause is correctly identified and the fix direction is sound.** Upstream `PostgresValueConverter.convertTimestampToLocalDateTime` uses `LocalDateTime.ofInstant(timestamp.toInstant(), ZoneOffset.systemDefault())`. For pre-epoch values, the legacy `java.util.Date`/`Calendar` builds the instant using the zone's historical **LMT (Local Mean Time)** offset, while `ofInstant(..., systemDefault())` inverts it with inconsistent rules — so `1900-01-01 00:00:00.123` under `Asia/Shanghai` is mis-parsed as `1900-01-01T00:05:43.123` (exactly matching FLINK-38835). Switching to `timestamp.toLocalDateTime()` reads the legacy wall-clock fields directly and avoids the instant↔zone round-trip; in my testing the 1900/1600/1969 boundaries all become correct under `Asia/Shanghai`.
- **The change is a pure improvement with no regression risk.** Post-epoch values (2020, 1970-01-01) are identical old vs. new in every zone; pre-epoch values are identical in fixed-offset zones (`UTC`/`GMT±hh:mm`/`Etc/GMT+8`); differences appear only in real regional zones with historical LMT (`Asia/Shanghai`/`Europe/Paris`), and always in the correcting direction. The fix only changes behavior where it was previously wrong.
- **±Infinity sentinel semantics are preserved.** The converter checks `POSITIVE_INFINITY_TIMESTAMP`/`NEGATIVE_INFINITY_TIMESTAMP` before calling `toLocalDateTime()`, so sentinels never go through the legacy field conversion — consistent with upstream.
- **Converter wiring is complete for both the snapshot and incremental paths.** `PostgresConnection` (default value converter / snapshot path) and `PostgresObjectUtils` (`newPostgresValueConverterBuilder` → `PostgresDialect` / `PostgresSourceFetchTaskContext`, WAL streaming path) both switch to `CustomPostgresValueConverter.of`; grep confirms no remaining `PostgresValueConverter.of(` call sites in main code.
- **Scope is well contained.** Only `convertTimestampToLocalDateTime` (for `timestamp` without tz) is overridden; `timestamptz` goes through `convertTimestampWithZone` → `OffsetDateTime` and is handled as an absolute-instant path — correctly left untouched here.
- **The `of()` factory faithfully mirrors upstream** (same `ZoneOffset.UTC`, `bigIntUnsignedMode=null`), with no config drift.

### Issues

#### Critical (must fix)

- **[PostgreSQLConnectorITCase.java + .github/workflows/utils.sh] The new assertion passes even without the fix, i.e. it provides no regression protection.** CI runs with `-Duser.timezone=$jvm_timezone`, and `random_timezone()` only produces **fixed-offset** `GMT±hh:mm` zones. In my testing, under `GMT+8:00`/`GMT-5:00`/`Etc/GMT+8`, `ofInstant(toInstant(), systemDefault())` and `toLocalDateTime()` return **identical** results (the bug does not reproduce); only real regional zones with LMT expose it. As a result the new `1900-01-01T00:00:00.123` assertion cannot distinguish "fixed vs. not fixed" in CI — a future Debezium upgrade or refactor that silently reverts to the instant-based logic would not be caught. Suggestion: pin the test JVM timezone to a regional zone with historical LMT (e.g. `Asia/Shanghai`), and add a **pure unit test** asserting pre-epoch results of `CustomPostgresValueConverter.convertTimestampToLocalDateTime` under a controlled `ZoneId` — a unit test needs no testcontainers and is the cheapest, most reliable regression guard.

#### Important (should fix)

- **[CustomPostgresValueConverter.java] Missing comment explaining why `toLocalDateTime()` is used instead of the instant conversion.** This is the most counter-intuitive line of the fix — `ofInstant(toInstant(), systemDefault())` looks "more canonical", yet it is exactly what breaks under LMT + pre-epoch. Without a comment a future maintainer may well-intentionedly revert it and reintroduce the bug. Suggestion: add a comment above `return timestamp.toLocalDateTime();` explaining the negative-epoch + historical-LMT offset issue and referencing FLINK-38835.
- **[PostgreSQLConnectorITCase.java] The assertion was relaxed from `isEqualTo(expected)` to `containsExactlyInAnyOrderElementsOf(expected)`, losing the ordering signal.** After introducing id=2 the row order is no longer deterministic, so the relaxation is understandable, but the original `isEqualTo` implicitly validated the snapshot(+I) → (-D) → incremental(+I) sequence. Suggestion: prefer `containsExactly` (ordered) with a deterministic update order, or assert per row grouped by op phase, rather than a whole-set in-any-order check.
- **[Test coverage] Pre-epoch boundary cases are insufficient.** Only `1900-01-01` is covered. Suggest adding at least: exactly `1970-01-01 00:00:00` (epoch boundary, verify no off-by-one), `1969-12-31 23:59:59.999999` (just pre-epoch + sub-second), a far-past value (e.g. the 1600s, where the LMT deviation is larger), and `±Infinity` (verify the sentinel path is not broken). These are cheap once the ZoneId is pinned in a unit test.

#### Minor (nice to have)

- **[CustomPostgresValueConverter.java] The constructor + `of()` is ~45 lines of boilerplate passthrough that will break at compile time when the parent constructor signature changes across Debezium upgrades.** The current approach (placing it in the `io.debezium.connector.postgresql` package to reach the `protected` parent constructor) is consistent with the existing `PostgresObjectUtils` hack and is acceptable; suggest noting in the class Javadoc that the signature must stay in sync with the `PostgresValueConverter` parent constructor.
- **[CustomPostgresValueConverter.java] The class Javadoc only says "correctly handles ... before 1970-01-01" and does not state that it applies only to `timestamp` (without tz), not `timestamptz`.** Suggest adding a scope note so readers don't assume timestamptz is also changed.
- **[column_type_test.sql] Missing trailing newline at end of file.** Add a final newline for consistency with the repo's other SQL fixtures.
- **[column_type_test.sql] The id=2 row duplicates all columns of id=1 except the time fields.** Readability is fine; if more boundary values are added later, consider parameterizing / multi-row INSERT to avoid fixture bloat.

### Verdict

**Ready to merge?** With fixes

**Reasoning:** The fix itself is correct, well-scoped, and in my testing has zero regressions for post-epoch and fixed-offset zones, only correcting values in the genuinely broken LMT + pre-epoch scenario. However, because CI (the merge gate) uses fixed-offset `GMT±hh:mm` zones, the new ITCase assertion would pass even without the fix — so the PR currently carries no effective regression protection. Before merging, please add a unit/IT regression case pinned to a real regional zone (e.g. `Asia/Shanghai`) and a comment explaining the `toLocalDateTime()` choice.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-postgres-cdc/src/test/java/org/apache/flink/cdc/connectors/postgres/table/PostgreSQLConnectorITCase.java [hunk 1]
```diff
@@ -689,18 +689,21 @@ void testAllTypes(boolean parallelismSnapshot) throws Throwable {
...
+                        "+I(1,[50],0,65535,2147483647,5.5,6.6,123.12345,404.4,true,Hello World,a,abc,abcd..xyz,2020-07-17T18:00:22.123,2020-07-17T18:00:22.123456,2020-07-17,18:00:22,500,{\"hexewkb\":\"0105000020e610000001000000010200000002000000a779c7293a2465400b462575025a46c0c66d3480b7fc6440c3d32b65195246c0\",\"srid\":4326},{\"hexewkb\":\"0101000020730c00001c7c613255de6540787aa52c435c42c0\",\"srid\":3187})",
+                        "+I(2,[50],0,65535,2147483647,5.5,6.6,123.12345,404.4,true,Hello World,a,abc,abcd..xyz,1900-01-01T00:00:00.123,1900-01-01T00:00:00.123456,1900-01-01,18:00:22,500,{\"hexewkb\":\"0105000020e610000001000000010200000002000000a779c7293a2465400b462575025a46c0c66d3480b7fc6440c3d32b65195246c0\",\"srid\":4326},{\"hexewkb\":\"0101000020730c00001c7c613255de6540787aa52c435c42c0\",\"srid\":3187})");
         List<String> actual = TestValuesTableFactory.getRawResultsAsStrings("sink");
-        Assertions.assertThat(actual).isEqualTo(expected);
+        Assertions.assertThat(actual).containsExactlyInAnyOrderElementsOf(expected);
```
[the remark this answers is not in the document]
Comment (lvyanquan) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-postgres-cdc/src/test/java/org/apache/flink/cdc/connectors/postgres/table/PostgreSQLConnectorITCase.java]: Thanks. I restored `testAllTypes` to its original ordered assertion and moved the pre-epoch regression coverage into a dedicated `testPreEpochTimestamp` ITCase. The new test filters the target row but verifies the exact `+I -> -D -> +I` sequence, so event ordering remains covered without relying on cross-row snapshot ordering.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-postgres-cdc/src/main/java/io/debezium/connector/postgresql/CustomPostgresValueConverter.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        final Timestamp timestamp = (Timestamp) data;
+
+        if (POSITIVE_INFINITY_TIMESTAMP.equals(timestamp)) {
+            return POSITIVE_INFINITY_LOCAL_DATE_TIME;
+        } else if (NEGATIVE_INFINITY_TIMESTAMP.equals(timestamp)) {
+            return NEGATIVE_INFINITY_LOCAL_DATE_TIME;
+        }
+
+        return timestamp.toLocalDateTime();
```
[the remark this answers is not in the document]
Comment (lvyanquan) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-postgres-cdc/src/main/java/io/debezium/connector/postgresql/CustomPostgresValueConverter.java:111]: Thanks. I added method-level Javadoc explaining that `column` and `fieldDefn` are intentionally unused because this override only changes how the JDBC `Timestamp` value is converted. Schema and field-specific handling remains in the parent converter flow.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-postgres-cdc/src/main/java/io/debezium/connector/postgresql/CustomPostgresValueConverter.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                null,
+                connectorConfig.includeUnknownDatatypes(),
+                typeRegistry,
+                connectorConfig.hStoreHandlingMode(),
+                connectorConfig.binaryHandlingMode(),
+                connectorConfig.intervalHandlingMode(),
+                connectorConfig.getUnavailableValuePlaceholder(),
+                connectorConfig.moneyFractionDigits());
+    }
+
```
[the remark this answers is not in the document]
Comment (lvyanquan) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-postgres-cdc/src/main/java/io/debezium/connector/postgresql/CustomPostgresValueConverter.java:84]: Addressed. The class and overridden method now document why the instant/system-default-zone round trip is incorrect for pre-epoch timestamps in regional time zones, why `Timestamp.toLocalDateTime()` preserves the wall-clock fields, and that the change applies only to PostgreSQL `timestamp` without time zone.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-postgres-cdc/src/main/java/io/debezium/connector/postgresql/CustomPostgresValueConverter.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        final Timestamp timestamp = (Timestamp) data;
+
+        if (POSITIVE_INFINITY_TIMESTAMP.equals(timestamp)) {
+            return POSITIVE_INFINITY_LOCAL_DATE_TIME;
+        } else if (NEGATIVE_INFINITY_TIMESTAMP.equals(timestamp)) {
+            return NEGATIVE_INFINITY_LOCAL_DATE_TIME;
+        }
+
+        return timestamp.toLocalDateTime();
+    }
```
[the remark this answers is not in the document]
Comment (lvyanquan) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-postgres-cdc/src/main/java/io/debezium/connector/postgresql/CustomPostgresValueConverter.java:112]: Addressed. I added `CustomPostgresValueConverterTest` covering null and non-`Timestamp` values, positive and negative infinity sentinels, timestamps in 1600 and 1900, immediately before and at the 1970 epoch boundary, and a post-epoch value under `Asia/Shanghai`. The test passes with both Flink 1.20 and Flink 2.2.

**Comment by lvyanquan** [comment]
Thanks @yuxiqian and @leonardBang. I updated the PR in commit `13681652c`.

Changes:

- Rebased the branch onto the latest `master` and resolved the conflicts.
- Added a dedicated `testPreEpochTimestamp` ITCase pinned to `Asia/Shanghai`. It verifies the `1900-01-01` value in both the snapshot and WAL update stages.
- Restored the original ordered assertion in `testAllTypes`; the new regression test verifies the target row’s exact `+I -> -D -> +I` sequence.
- Added `CustomPostgresValueConverterTest` covering null and non-timestamp input, positive and negative infinity, 1600, 1900, the pre-epoch and epoch boundaries, and post-epoch timestamps.
- Added class and method documentation explaining the historical LMT issue, the reason for using `Timestamp.toLocalDateTime()`, the intentionally unused parameters, and the scope of the change.
- Moved the pre-epoch test data setup into the dedicated ITCase, so `column_type_test.sql` is no longer changed and the trailing-newline/duplicated-fixture observations no longer apply to this diff.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-postgres-cdc/src/main/java/io/debezium/connector/postgresql/CustomPostgresValueConverter.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        final Timestamp timestamp = (Timestamp) data;
+
+        if (POSITIVE_INFINITY_TIMESTAMP.equals(timestamp)) {
+            return POSITIVE_INFINITY_LOCAL_DATE_TIME;
+        } else if (NEGATIVE_INFINITY_TIMESTAMP.equals(timestamp)) {
+            return NEGATIVE_INFINITY_LOCAL_DATE_TIME;
+        }
+
+        // Preserve the timestamp's wall-clock fields instead of applying historical zone offsets.
+        return timestamp.toLocalDateTime();
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-postgres-cdc/src/main/java/io/debezium/connector/postgresql/CustomPostgresValueConverter.java]: Thanks for the update — overall the change looks good to me. My only concern is upgrade compatibility. Comparing the old and new implementations under Asia/Shanghai, Europe/Paris, UTC and GMT+08:00, only `timestamp` (without time zone) values before 1970-01-01 under regional time zones change, and checkpoint state is not affected — but for a running job this change is silent. Users who have already worked around the historical offset downstream (e.g. compensated via a transform/UDF, or ran a correction job) would get double-corrected after a routine upgrade, i.e. data that is currently correct would start producing wrong values. For a bug fix shipped in a minor release, I'd lean towards not changing the existing default behavior.

Since such scenarios are also quite rare, would it make sense to introduce a config option that keeps the current behavior by default and lets users opt in deliberately? Naming-wise, how about: `scan.pre-epoch-timestamp.wall-clock-conversion.enabled = false`

Comment (lvyanquan) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-postgres-cdc/src/main/java/io/debezium/connector/postgresql/CustomPostgresValueConverter.java]: Good point, agreed — a fix in a minor release shouldn't silently change the output of jobs that already work. Added your option `scan.pre-epoch-timestamp.wall-clock-conversion.enabled` (default `false` = previous behavior kept, `true` = the wall clock stored in PostgreSQL is kept).

It works in the SQL source, the DataStream source (same key in `debeziumProperties`) and the Postgres pipeline connector; en/zh docs updated. A unit test asserts that the default path still returns exactly what a stock `PostgresValueConverter` returns, and ITCases pin both modes to `Asia/Shanghai`.

One limitation, now documented: the option only affects the incremental snapshot source. With `scan.incremental.snapshot.enabled=false` the record values are converted by Debezium's own `PostgresConnectorTask`, so the option is a no-op there — covering that would require shadowing more Debezium classes. Happy to include it here if you prefer.

**Comment by lvyanquan** [comment]
Pushed `7c52a4f39` for the upgrade-compatibility concern:

- New option `scan.pre-epoch-timestamp.wall-clock-conversion.enabled`, default `false` -> previous behavior is kept (`super.convertTimestampToLocalDateTime`); `true` -> the wall clock stored in PostgreSQL is kept.
- Supported by the SQL source, the DataStream source (same key in `debeziumProperties`) and the Postgres pipeline connector; en/zh docs updated for both connectors.
- Tests: converter unit tests for both modes (including equality with a stock `PostgresValueConverter`), an opt-in ITCase covering the snapshot and WAL phases, an ITCase asserting the unchanged default behavior, and a pipeline factory wiring test. Green on Flink 1.20 and Flink 2.2 (JDK 17).
- Documented limitation: no effect when `scan.incremental.snapshot.enabled=false`.

Supporting data point: under the previous conversion the same pre-epoch row is read as `1900-01-01T00:05:43.123` during the snapshot phase but as `1900-01-01T00:00:00.123` from the WAL (`Asia/Shanghai`), so the two phases of one job disagree. With the option enabled both phases return the stored value.

**Review by leonardBang — APPROVED** [review]
Thanks @lvyanquan for the update, +1 from my side
